address/address.go

```diff
@@ -47,327 +77,236 @@ const (
+func ValidateChecksum(ingest, expect []byte) bool {
+	digest := Checksum(ingest)
+	return bytes.Equal(digest, expect)
 }
 
-// --
-// TODO: find a better place for the things below
+func addressHash(ingest []byte) []byte {
+	return hash(ingest, payloadHashConfig)
+}
 
-// encode encodes hrp(human-readable part) a version(byte) and data(32bit data array)
-func encode(hrp string, version byte, data []byte) (string, error) {
-	if len(hrp) != 2 {
-		return "", fmt.Errorf("hrp has invalid length: hrp=%d", len(hrp))
+func newAddress(protocol Protocol, payload []byte) Address {
+	if protocol < ID || protocol > BLS {
+		panic("invalid protocol")
 	}
-	if len(data) != HashLength {
-		return "", fmt.Errorf("data is malformed: data length=%d", len(data))
+	explen := 1 + len(payload)
+	buf := make([]byte, explen)
+
+	buf[0] = protocol
+	if c := copy(buf[1:], payload); c != len(payload) {
+		panic("copy data length is inconsistent")
 	}
-	for p, c := range hrp {
-		if c < 33 || c > 126 {
-			return "", fmt.Errorf("invalid character human-readable part : hrp[%d]=%d", p, c)
-		}
+	log.Debugf("new address protocol: %x, payload: %v", protocol, payload)
+	return Address{string(buf)}
+}
+
+func encode(network Network, addr Address) (string, error) {
+	// TODO(frrist): not sure if this is the correct thing to do here
+	if addr == Undef {
+		return "", nil
 	}
-	if version > 31 {
-		return "", fmt.Errorf("version too large")
+	var ntwk string
+	switch network {
+	case Mainnet:
+		ntwk = MainnetPrefix
+	case Testnet:
+		ntwk = TestnetPrefix
+	default:
+		panic("invalid network byte")
 	}
 
-	hrp = strings.ToLower(hrp)
-	combined := append([]byte{version}, Base32.EncodeToBytes(data)...)
-	sum := createChecksum(hrp, combined)
-
-	toEncode := append(combined, sum...)
-	encoded := make([]byte, len(toEncode))
-
-	for i, p := range toEncode {
-		if p >= 32 {
-			return "", fmt.Errorf("invalid data")
-		}
-		encoded[i] += Base32Charset[p]
+	var strAddr string
+	switch addr.Protocol() {
+	case SECP256K1, Actor, BLS:
+		cksm := Checksum(append([]byte{addr.Protocol()}, addr.Payload()...))
+		strAddr = ntwk + fmt.Sprintf("%d", addr.Protocol()) + AddressEncoding.WithPadding(-1).EncodeToString(append(addr.Payload(), cksm[:]...))
+	case ID:
+		strAddr = ntwk + fmt.Sprintf("%d", addr.Protocol()) + fmt.Sprintf("%d", leb128.ToUInt64(addr.Payload()))
+	default:
+		return "", errors.New("invalid protocol byte")
 	}
-
-	return hrp + string(encoded), nil
+	log.Debugf("encoded address: %s", strAddr)
+	return strAddr, nil
 }
 
-// decode decodes a filecoin address and returns
-// the hrp(human-readable part), version (1byte) and data(32bit data array)
-// or an error.
-func decode(addr string) (string, byte, []byte, error) {
-	if len(addr) < 2 {
-		return "", 0, nil, fmt.Errorf("address too short")
-	}
-	if len(addr) > 41 {
-		return "", 0, nil, fmt.Errorf("too long: len=%d", len(addr))
+func decode(a string) (Address, error) {
```

Comment:
should also have a check for max length.

Reply:
addressed here: https://github.com/filecoin-project/go-filecoin/pull/2186/files#diff-2026d132a1aa6ec72a1af72e5763011cR272